File: J9Ch/src/J_8_Data_Structure/Related_7/python_cpp/_540_Zigzag_Iterator.cpp

class ZigzagIterator {
private:
    int p;
    vector<int>::iterator begin[2], end[2];

public:
    /**
     * @param v1 v2 two 1d vectors
     */
    ZigzagIterator(vector<int>& v1, vector<int>& v2) {
        // initialize your data structure here.
        begin[0] = v1.begin(), begin[1] = v2.begin();
        end[0] = v1.end(), end[1] = v2.end();
        p = 0;
    }

    int next() {
        // Write your code here
        int elem;
        if (begin[0] == end[0])
            elem = *begin[1]++;
        else if (begin[1] == end[1])
            elem = *begin[0]++;
        else {
            elem = *begin[p]++;
            p = (p + 1) % 2;
        }
        return elem;
    }

    bool hasNext() {
        // Write your code here
        return begin[0] != end[0] || begin[1] != end[1];
    }
};

/**
 * Your ZigzagIterator object will be instantiated and called as such:
 * ZigzagIterator solution(v1, v2);
 * while (solution.hasNext()) result.push_back(solution.next());
 * Ouptut result
 */

File: J9Ch/src/J_8_DS/other/Iterator/_540_Zigzag_Iterator.java
package J_8_DS.other.Iterator;

import java.util.Iterator;
import java.util.List;

/**540 Zigzag Iterator

 * Created by tianhuizhu on 7/10/17.
 */

/**
 * Your ZigzagIterator object will be instantiated and called as such:
 * ZigzagIterator solution = new ZigzagIterator(v1, v2);
 * while (solution.hasNext()) result.add(solution.next());
 * Output result
 */

public class _540_Zigzag_Iterator {

    public class ZigzagIterator {

        public Iterator<Integer> it1;
        public Iterator<Integer> it2;
        public int turns;

        /**
         * @param v1 v2 two 1d vectors
         */
        public ZigzagIterator(List<Integer> v1, List<Integer> v2) {
            // initialize your data structure here.
            this.it1 = v1.iterator();
            this.it2 = v2.iterator();
            turns = 0;
        }

        public int next() {
            // Write your code here
            turns++;
            if((turns % 2 == 1 && it1.hasNext()) || (!it2.hasNext())) {
                return it1.next();
            } else if((turns % 2 == 0 && it2.hasNext()) || (!it1.hasNext())) {
                return it2.next();
            }
            return -1;
        }

        public boolean hasNext() {
            // Write your code here
            return it1.hasNext() || it2.hasNext();
        }
    }


////////////////////////////////////////////////////////////////////////////////





////////////////////////////////////////////////////////////////////////////////





////////////////////////////////////////////////////////////////////////////////





////////////////////////////////////////////////////////////////////////////////





////////////////////////////////////////////////////////////////////////////////






}